src/python/pants/releases/reversion.py

```diff
@@ -0,0 +1,146 @@
+  versions, and having a version in its filename.
+  """
+  mutated_files = set()
+  dst_record_file = None
+  for src, dst in mutated_file_tuples:
+    if src == src_record_file:
+      dst_record_file = dst
+    else:
+      mutated_files.add(dst)
+  if not dst_record_file:
+    raise Exception('Malformed whl: `{}` was not rewritten.'.format(src_record_file))
+
+  output_records = []
+  for line in read_file(os.path.join(workspace, dst_record_file)).strip().split('\n'):
+    filename, fingerprint_str, size_str = line.strip().rsplit(',', 3)
+    if filename in mutated_files:
+      fingerprint_str, size_str = fingerprint_file(workspace, filename)
+      output_records.append(','.join((filename, fingerprint_str, size_str)))
+    else:
+      output_records.append(line)
+
+  safe_file_dump(os.path.join(workspace, dst_record_file), '\r\n'.join(output_records) + '\r\n')
+
+
+def reversion(args):
+  with temporary_dir() as workspace:
+    # Extract the input.
+    with open_zip(args.whl_file, 'r') as whl:
+      src_filenames = whl.namelist()
+      whl.extractall(workspace)
+
+    # Determine the location of the `dist-info` directory.
+    dist_info_dir = locate_dist_info_dir(workspace)
+    record_file = os.path.join(dist_info_dir, 'RECORD')
+
+    # Load metadata for the input whl.
+    with open(os.path.join(workspace, dist_info_dir, 'metadata.json'), 'r') as info:
+      metadata = json.load(info)
+    input_version = metadata['version']
+
+    # Rewrite and move all files (including the RECORD file), recording which files need to be
+    # re-fingerprinted due to content changes.
+    dst_filenames = []
+    refingerprint = []
+    for src_filename in src_filenames:
+      if os.path.isdir(os.path.join(workspace, src_filename)):
+        continue
+      dst_filename = replace_in_file(workspace, src_filename, input_version, args.target_version)
+      if dst_filename is not None:
+        refingerprint.append((src_filename, dst_filename))
+      else:
+        dst_filename = src_filename
+      dst_filenames.append(dst_filename)
+
+    # Refingerprint relevant entries in the RECORD file under their new names.
+    rewrite_record_file(workspace, record_file, refingerprint)
+
+    # Create a new output whl in the destination.
+    dst_whl_filename = os.path.basename(args.whl_file).replace(input_version, args.target_version)
+    dst_whl_file = os.path.join(args.dest_dir, dst_whl_filename)
+    with open_zip(dst_whl_file, 'w', zipfile.ZIP_DEFLATED) as whl:
+      for dst_filename in dst_filenames:
+        whl.write(os.path.join(workspace, dst_filename), dst_filename)
+
+    print('Wrote whl with version {} to {}.\n'.format(args.target_version, dst_whl_file))
+
+
+def main():
+  """Given an input whl file and target version, create a copy of the whl with that version.
+
+  This is accomplished via a ton of string replacement: if the input version is not sufficiently
```

Comment:
IIUC there is exactly 1 wheel we need to do a find-replace on non-wheel metadata files ("content files"), that for the `pantsbuild.pants` wheel itself. Pants plugin wheel won't have a this version embedded in "content files" and all other dependencies won't be re-versioned at all. If this is correct, we know the 1 file to edit in that case. Would it be reasonable to just have the tool take an optional list of positional arguments that correspond to wheel "content files" to edit or even positional args forming a command line to execute in the chroot of the exploded wheel for extra transformations needed?

All this said, I think this is probably fine as-is.